mantidimaging/core/utility/leak_tracker.py

```diff
@@ -0,0 +1,113 @@
+# Copyright (C) 2022 ISIS Rutherford Appleton Laboratory UKRI
+# SPDX - License - Identifier: GPL-3.0-or-later
+import gc
+import sys
+import traceback
+import weakref
+from typing import NamedTuple, List, Set, Iterable
+
+from numpy import ndarray
+
+
+class ItemInfo(NamedTuple):
+    ref: weakref.ref
+    msg: str
+    created: List[str]
+
+
+def obj_to_string(obj, relative=None) -> str:
+    """Shows object type, id and extra info for some classes"""
+    extra_info = ""
+    if isinstance(obj, dict) and relative is not None:
+        keys = {k for k, v in obj.items() if v is relative}
+        extra_info += f" keys={keys}"
+
+    try:
+        if hasattr(obj, "name"):
+            extra_info += f" obj.name={obj.name}"
+
+        if hasattr(obj, "id"):
+            extra_info += f" obj.id={obj.id}"
+    except RuntimeError as e:
+        # Can happen with a deleted QWidget
+        extra_info += str(e)
+
+    if isinstance(obj, ndarray):
+        extra_info += f" ndarray: {obj.shape} {obj.dtype}"
+
+    return f"{type(obj)} pyid={id(obj)} {extra_info}"
+
+
+def find_owners(obj, depth: int, path: List[str] = None, ignore: Set[int] = None) -> List[List[str]]:
+    """Recursively track though references to objects and return a list of routes"""
+    all_routes = []
+    if path is None:
+        path = [obj_to_string(obj)]
+    if ignore is None:
+        ignore = set()
+    ignore.add(id(sys._getframe()))
+    ignore.add(id(path))
+    ignore.add(id(obj))
+
+    new_refs = [r for r in gc.get_referrers(obj) if id(r) not in ignore]
+    ignore.add(id(new_refs))
+    if len(new_refs) == 0:
+        all_routes.append(path)
+    for owner in new_refs:
+        if type(owner).__name__ in ['frame', 'function', 'list_iterator']:
+            continue
+        route = [obj_to_string(owner, obj)] + path
+        ignore.add(id(route))
+        if depth > 0:
+            new_routes = find_owners(owner, depth - 1, route, ignore)
+            for new_route in new_routes:
+                all_routes.append(new_route)
```

Comment:
Would this also work if we used `all_routes.extend(new_routes)` instead of the `for` loop?